lib/smart_answer_flows/maternity-paternity-calculator/maternity_calculator_flow.rb

```diff
@@ -277,68 +277,7 @@ def define
         end
 
         ## Maternity outcomes
-        outcome :maternity_leave_and_pay_result do
-          precalculate :pay_method do
-            calculator.pay_method = (
-              if monthly_pay_method
-                if monthly_pay_method == "specific_date_each_month" && pay_day_in_month > 28
-                  "last_day_of_the_month"
-                else
-                  monthly_pay_method
-                end
-              elsif smp_calculation_method == "weekly_starting"
-                smp_calculation_method
-              elsif calculator.pay_pattern
-                calculator.pay_pattern
-              end
-            )
-          end
-          precalculate :smp_a do
-            sprintf("%.2f", calculator.statutory_maternity_rate_a)
-          end
-          precalculate :smp_b do
-            sprintf("%.2f", calculator.statutory_maternity_rate_b)
-          end
-          precalculate :lower_earning_limit do
-            sprintf("%.2f", calculator.lower_earning_limit)
-          end
-
-          precalculate :notice_request_pay do
-            calculator.notice_request_pay
-          end
-
-          precalculate :below_threshold do
-            calculator.earnings_for_pay_period &&
-              calculator.average_weekly_earnings_under_lower_earning_limit?
-          end
-
-          precalculate :not_entitled_to_pay_reason do
-            if below_threshold
-              :must_earn_over_threshold
-            else
-              not_entitled_to_pay_reason
-            end
-          end
-
-          precalculate :total_smp do
-            if not_entitled_to_pay_reason.blank?
-              sprintf("%.2f", calculator.total_statutory_pay)
-            end
-          end
-
-          precalculate :pay_dates_and_pay do
-            if not_entitled_to_pay_reason.blank?
-              lines = calculator.paydates_and_pay.map do |date_and_pay|
-                %(#{date_and_pay[:date].strftime('%e %B %Y')}|£#{sprintf('%.2f', date_and_pay[:pay])})
-              end
-              lines.join("\n")
-            end
```

Comment:
This isn't a very nice bit of code. The conditional statement should be in the statement calling `pay_dates_and_pay` and not within the method itself. It would then be a lot easier to tie the conditional to the requirement context. I didn't fix that because I thought it was taking the refactoring too far and ....

If you look at `not_entitled_to_pay_reason` it is populated in two ways:

1. Directly via `not_entitled_to_pay_reason =`
2. As `:must_earn_over_threshold` if `pay_below_threshold?`

If you compare `above_lower_earning_limit` with `pay_below_threshold?` one is the opposite of the other:

```ruby
    def average_weekly_earnings_under_lower_earning_limit?
      average_weekly_earnings < lower_earning_limit
    end

    def above_lower_earning_limit
      average_weekly_earnings > lower_earning_limit
    end
```

So in effect the two version of `pay_dates_and_pay` behave the same way. 

The only exception is if the `not_entitled_to_pay_reason` has been defined directly - and if that is the case the user will be kicked out of the flow before the point where `pay_dates_and_pay` needs to be processed.

There is one edge case - when `average_weekly_earnings == lower_earning_limit` but that highlights an inconsistency between the way `pay_dates_and_pay` is calculated here, and how it is in the rest of the flow. Also the edge case is very small because `average_weekly_earnings` is calculated and stored to many decimal places.

Finally, if you don't do it this way, you'll have to define `pay_dates_and_pay` differently in each of the calculators.

Reply:
Ah wait I don't think we even use this precalculated method. All the templates have explicit references to "calculator. pay_dates_and_pay" anyway.